deployer/grafana/grafana_tokens.py

```diff
@@ -0,0 +1,210 @@
+            f"Error was {response.text}."
+        )
+        response.raise_for_status()
+
+    service_account_no = response.json()["totalCount"]
+    if service_account_no == 0:
+        return
+
+    service_accounts = response.json()["serviceAccounts"]
+    return next((sa["id"] for sa in service_accounts if sa["name"] == "deployer"), None)
+
+
+def get_deployer_token(sa_endpoint, sa_id, headers):
+    """
+    Iterate the list of tokens created for the `deployer` service account and
+    retrieve the token with the name `deployer`.
+
+    Returns:
+        token, if deployer token exists
+        None otherwise
+    """
+    response = requests.get(
+        f"{sa_endpoint}/{sa_id}/tokens",
+        headers=headers,
+    )
+    if not response.ok:
+        print(
+            f"An error occured when retrieving the tokens the service account with id {sa_id}.\n"
+            f"Error was {response.text}."
+        )
+        response.raise_for_status()
+
+    for token in response.json():
+        if token["name"] == "deployer":
+            return token
+
+    return
+
+
+def create_deployer_token(sa_endpoint, sa_id, headers):
+    token_request_body = {"name": "deployer", "role": "Admin"}
+    response = requests.post(
+        f"{sa_endpoint}/{sa_id}/tokens",
+        data=json.dumps(token_request_body),
+        headers=headers,
+    )
+
+    if not response.ok:
+        print(
+            "An error occured when creating the token for the deployer service account.\n"
+            f"Error was {response.text}."
+        )
+        response.raise_for_status()
+
+    print_colour("Successfully generated the `deployer` token!")
+
+    return response.json()["key"]
+
+
+@app.command()
+def new_grafana_token(
+    cluster=typer.Argument(
+        ...,
+        help="Name of cluster for who's Grafana deployment to generate a new deployer token",
+    )
+):
+    """
+    Generate an API token for the cluster's Grafana `deployer` service account
+    and store it encrypted inside a `enc-grafana-token.secret.yaml` file.
+    """
+    grafana_host = get_grafana_url(cluster)
+    sa_endpoint = f"https://{grafana_host}/api/serviceaccounts"
+    headers = build_service_account_request_headers()
+
+    overwrite = False
+    token = None
+
+    # Check if there's a service account called "deployer" already
+    sa_id = get_deployer_service_account_id(sa_endpoint, headers)
+    if not sa_id:
+        print_colour("Creating a `deployer` service account...", "yellow")
+        sa_id = create_deployer_service_account(sa_endpoint, headers)
+    else:
+        # Check if there's a "deployer" token for the "deployer service account" already
+        token = get_deployer_token(sa_endpoint, sa_id, headers)
+
+    if token:
+        print_colour(
+            "Token already exists!\n"
+            f'Checking if has expired: {token["hasExpired"]}\n',
+            "yellow",
+        )
+        print_colour(
```

Comment:
Is there a reason we would want to keep an expired token? Could we automatically replace them if it exists and is expired?

Reply:
Yes, no reason to keep them.
This script creates tokens with no expiration date however. I couldn't find in their docs how to set one https://grafana.com/docs/grafana/latest/developers/http_api/serviceaccount